src/main/scala/is/hail/utils/SetupIterator.scala

```diff
@@ -0,0 +1,20 @@
+package is.hail.utils
+
+class SetupIterator[T](it: Iterator[T], setup: () => Unit) extends Iterator[T] {
+  private[this] var needsSetup = true
+
+  def hasNext: Boolean = {
+    if (needsSetup) {
```

Comment:
why is it necessary for hasNext to run `setup` as well?

Reply:
I'm pretty sure we have iterators that use `hasNext` to set up the value that will be returned in `next`. If I always `reset` in `next` then I'll destroy the value before it's even returned.

Pictorially what I want to achieve is this:

```
| value | value | value | ...
```

wherein the pipe's represent me resetting the context and the values are valid for the time period between two resets. The first reset is not necessary, but it seemed easier to call reset before every value.

The real trouble is this sequence of operations:
```
next()
hasNext()
val v2 = next()
```
When `hasNext` is called, Hail-style iterators will try to read a value from their sources, storing it in the region if a value exists. Unfortunately, if I `clear` before each call to `next`, then I'll destroy that value. I could also just clear in `hasNext`, but sometimes we _know_ there must be two values so we:

```
val rv = next().copy()
val rv2 = next()
```
and that puts unnecessary garbage in the region (rv).